utilities/backupable/backupable_db.cc

```diff
@@ -2159,84 +2215,182 @@ Status BackupEngineImpl::BackupMeta::LoadFromFile(
 
-    uint64_t size;
+    uint64_t actual_size;
     const std::shared_ptr<FileInfo> file_info = GetFile(filename);
     if (file_info) {
-      size = file_info->size;
+      actual_size = file_info->size;
     } else {
       std::string abs_path = backup_dir + "/" + filename;
       auto e = abs_path_to_size.find(abs_path);
       if (e == abs_path_to_size.end()) {
         return Status::Corruption("Pathname in meta file not found on disk: " +
                                   abs_path);
       }
-      size = e->second;
+      actual_size = e->second;
     }
 
-    if (components.size() < 3) {
-      return Status::Corruption("File checksum is missing for " + filename +
-                                " in " + meta_filename_);
+    if (schema_major_version >= 2) {
+      if (components.size() % 2 != 1) {
+        return Status::Corruption(
+            "Bad number of line components for file entry.");
+      }
+    } else {
+      if (components.size() < 3) {
+        return Status::Corruption("File checksum is missing for " + filename +
+                                  " in " + meta_filename_);
+      }
+      // WART: The checksums are crc32c, not original crc32
+      if (components[1] != "crc32") {
+        return Status::Corruption("Unknown checksum type for " + filename +
+                                  " in " + meta_filename_);
+      }
+      if (components.size() > 3) {
+        return Status::Corruption("Extra data for entry " + filename + " in " +
+                                  meta_filename_);
+      }
     }
 
-    // WART: The checksums are crc32c, not original crc32
-    if (components[1] != "crc32") {
-      return Status::Corruption("Unknown checksum type for " + filename +
-                                " in " + meta_filename_);
+    std::string checksum_hex;
+    for (unsigned i = 1; i < components.size(); i += 2) {
+      const std::string& field_name = components[i];
+      const std::string& field_data = components[i + 1];
+
+      // WART: The checksums are crc32c, not original crc32
+      if (field_name == "crc32") {
+        uint32_t checksum_value =
+            static_cast<uint32_t>(strtoul(field_data.c_str(), nullptr, 10));
+        if (components[2] != ROCKSDB_NAMESPACE::ToString(checksum_value)) {
+          return Status::Corruption("Invalid checksum value for " + filename +
+                                    " in " + meta_filename_);
+        }
+        checksum_hex = ChecksumInt32ToHex(checksum_value);
+      } else if (field_name == "size") {
+        uint64_t ex_size =
+            std::strtoull(field_data.c_str(), nullptr, /*base*/ 10);
+        if (ex_size != actual_size) {
+          return Status::Corruption("For file " + filename + " expected size " +
+                                    ToString(ex_size) + " but found size" +
+                                    ToString(actual_size));
+        }
+      } else if (StartsWith(field_name, "ni_")) {
```

Comment:
I am not sure "ni_" is the best choice.  It seems a little opaque, but I guess this is internal.  If you want "ni_", would something like "ni::" be better, as it uses a (potentially) clearer delimiter

Reply:
Sounds reasonable.